app/View/Elements/serverRuleElements/pull.ctp

```diff
@@ -1,80 +1,186 @@
+        </div>
+        <div style="display: flex;">
+            <textarea style="width:100%;" placeholder='{"timestamp": "30d"}' type="text" value="" id="urlParams" required="required" data-original-title="" title="" rows="3"
+            ><?= !empty($ruleUrlParams) ? json_encode(h($ruleUrlParams), JSON_PRETTY_PRINT) : '' ?></textarea>
         </div>
     </div>
 </div>
+
+<?php
+echo $this->element('genericElements/assetLoader', array(
+    'js' => array(
+        'codemirror/codemirror',
+        'codemirror/modes/javascript',
+        'codemirror/addons/closebrackets',
+        'codemirror/addons/lint',
+        'codemirror/addons/jsonlint',
+        'codemirror/addons/json-lint',
+    ),
+    'css' => array(
+        'codemirror',
+        'codemirror/show-hint',
+        'codemirror/lint',
+    )
+));
+?>
+
+<script>
+var cm;
+$(function() {
+    var serverID = "<?= isset($id) ? $id : '' ?>"
+    <?php if ($context == 'servers'): ?>
+    addPullFilteringRulesToPicker()
+    <?php endif; ?>
+    setupCodeMirror()
+
+    function addPullFilteringRulesToPicker() {
+        var $rootContainer = $('div.server-rule-container-pull')
+        var $pickerTags = $rootContainer.find('select.rules-select-picker-tag')
+        var $pickerOrgs = $rootContainer.find('select.rules-select-picker-org')
+        if (serverID !== "") {
+            $pickerOrgs.parent().children().prop('disabled', true)
+            $pickerTags.parent().children().prop('disabled', true)
+            getPullFilteringRules(
+                function(data) {
+                    addOptions($pickerTags, data.tags)
+                    addOptions($pickerOrgs, data.organisations)
+                    $('div.notice-pull-rule-fetched.alert-success').show()
+                },
+                function(errorMessage) {
+                    showMessage('fail', '<?= __('Could not fetch remote sync filtering rules.') ?>');
+                    $('div.notice-pull-rule-fetched.alert-warning').show().find('.reason').text(errorMessage)
+                    $pickerTags.parent().remove()
+                    $pickerOrgs.parent().remove()
+                },
+                function() {
+                    $('div.notice-pull-rule-fetched.alert-primary').hide()
+                    $pickerTags.parent().children().prop('disabled', false).trigger('chosen:updated')
+                    $pickerOrgs.parent().children().prop('disabled', false).trigger('chosen:updated')
+                },
+            )
+        } else {
+            $('div.notice-pull-rule-fetched.alert-warning').show().find('.reason').text('<?= __('The server must first be saved in order to fetch remote synchronisation rules.') ?>')
+            $pickerTags.parent().remove()
+            $pickerOrgs.parent().remove()
+            $('div.notice-pull-rule-fetched.alert-primary').hide()
+        }
+    }
+
+    function getPullFilteringRules(cb, fcb, acb) {
```

Comment:
For maintainability, probably better to give the anonymous functions names that are a bit more clear as to what their function is.